### PR TITLE
feat(messaging): add support for setDeliveryMetricsExportToBigQuery

### PR DESCRIPTION
**Pull request by pavan168**
### Description

We have had very specific requirements to send out push notifications using FCM and have implemented an in-house solution using the Firebase Admin SDK. However, we don't get access to the message delivery information in the Firebase Console anymore and want to use BigQuery to analyse this information.

Reading through [the documentation](https://firebase.google.com/docs/cloud-messaging/understand-delivery?platform=android#enable-message-delivery-data-export) we found that the message delivery data export needs to be enabled on the device (after asking for consent of course) and both [ios](https://firebase.google.com/docs/reference/ios/firebasemessaging/api/reference/Classes/FIRMessagingExtensionHelper#-exportdeliverymetricstobigquerywithmessageinfo:) and [android](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/FirebaseMessaging#public-method-summary) native methods are not exposed as per the RNFirebase documentation. 

With this PR, we can use `setDeliveryMetricsExportToBigQuery` method to pass down the status (a boolean value) and have used `deliveryMetricsExportToBigQueryEnabled()` method that exist already on android to return the updated status whilst iOS does not have a similar method, `isDeliveryMetricsExportToBigQueryEnabled` is introduced that holds the most up-to-date status that can be used as per your requirement.

### Related issues
n/a

### Release Summary
n/a

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes :fire:
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [x] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

1. iOS test cases
ios

2. Android test cases
android

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

**Comment by mikehardy** [comment]
Hey! Thanks for your patience on this - I'm still catching up (as fast as I can, but never fast enough...) after summer travels. Queued for review

File: packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingModule.java [hunk 1]
```diff
@@ -221,10 +221,29 @@ public void unsubscribeFromTopic(String topic, Promise promise) {
...
+                rejectPromiseWithExceptionMap(promise, task.getException());
+              }
+            });
+  }
+
   @Override
   public Map<String, Object> getConstants() {
     final Map<String, Object> constants = new HashMap<>();
     constants.put("isAutoInitEnabled", FirebaseMessaging.getInstance().isAutoInitEnabled());
+    constants.put("isDeliveryMetricsExportToBigQueryEnabled", FirebaseMessaging.getInstance().deliveryMetricsExportToBigQueryEnabled());
```
Comment (mikehardy) [inline packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingModule.java:246]: Interesting - I was under the impression that constants were fetched once and cached such that setting new values into constants would not be reflected in future checks, but your test probes this, so, :+1:

**Review by mikehardy — APPROVED** [review]
This looks great. For a little bit I was confused why the iOS implementation wasn't actually *doing* anything, then I remembered our previous conversation where it came out that on the iOS side you can only send things per-message + on receipt. At which point the property is perfect as it may be queried by an extension

Great work here, all around

**Comment by mikehardy** [comment]
v15.6.0 just went out with this feature - good luck with your project!

File: packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingModule.java (hunk 1 above)
Comment (pavan168) [reply packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingModule.java:246]: I had the same thought at first. Gotta say that the tests were super helpful for a starter 👍

Comment (pavan168) [reply packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingModule.java:246]: Thank you very much @mikehardy Hoping to contribute more in future!

**Comment by mattThousand** [comment]
@mikehardy @pavan168 in order for an iOS app to export message data to BigQuery, I will still need to make my own extension, is that correct? But on Android, all I would need to do is call `setDeliveryMetricsExportToBigQuery(true)` from the javascript layer?

Or do the changes introduced in this PR mean that we no longer need to make our own Notification Service Extension? 

I suppose I am a bit confused as to what the iOS changes introduced in this PR do, exactly 🤔

**Comment by mikehardy** [comment]
I don't believe there is a way to avoid the extension on iOS bec you have to send it every notification. The PR made the status of user acceptance very accessible, but it still requires an extension. If this isn't clear in the docs you can hit the edit button at the top of any page and drop a note in to help future intrepid explorers 😁

**Pull request merged**